README.md
# WP Post
Simple WP Post Class With Advanced Options.

[![Latest Stable Version](https://poser.pugx.org/varunsridharan/wp-post/version)](https://packagist.org/packages/varunsridharan/wp-post)
[![Total Downloads](https://poser.pugx.org/varunsridharan/wp-post/downloads)](https://packagist.org/packages/varunsridharan/wp-post)
[![Latest Unstable Version](https://poser.pugx.org/varunsridharan/wp-post/v/unstable)](//packagist.org/packages/varunsridharan/wp-post)
[![License](https://poser.pugx.org/varunsridharan/wp-post/license)](https://packagist.org/packages/varunsridharan/wp-post)
[![composer.lock available](https://poser.pugx.org/varunsridharan/wp-post/composerlock)](https://packagist.org/packages/varunsridharan/wp-post)
 

## Installation
The preferred way to install this extension is through [Composer](http://getcomposer.org/download/).

To install **WP_Post library**, simply:

    $ composer require varunsridharan/wp-post

The previous command will only install the necessary files, if you prefer to **download the entire source code** you can use:

    $ composer require varunsridharan/wp-post --prefer-source

You can also **clone the complete repository** with Git:

    $ git clone https://github.com/varunsridharan/wp-post.git

Or **install it manually**:

[Download WP_Post.php](https://raw.githubusercontent.com/varunsridharan/wp-post/master/class-post.php):

    $ wget https://raw.githubusercontent.com/varunsridharan/wp-post/master/class-post.php

## Usage
```php
$post_data = new \Varunsridharan\WordPress\Post(22);
```

### Get Post Title
```php
echo $post_data->title();
```

### Get Featured Image URL
```php
if($post_data->has_featured_image()){
    echo '<img src="'.$post_data->featured_image_url('large').'"/>';
}
```

## Available Class Methods

* `has_featured_image`
* `featured_image_id`
* `featured_image_url`
* `permalink`
* `author`
* `has_content`
* `has_excerpt`
* `title`
* `slug`
* `parent`
* `type`
* `page_template`
* `excerpt`
* `content`
* `status`
* `is_status`
* `get_meta`
* `update_meta`
* `add_meta`
* `delete_meta`
* `taxonomies`
* `get_terms`
* `set_terms`
* `post`
* `id`


---

<!-- START common-footer.mustache  -->
## 📝 Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

[Checkout CHANGELOG.md](https://github.com/varunsridharan/wp-post/blob/main/CHANGELOG.md)


## 🤝 Contributing
If you would like to help, please take a look at the list of [issues](https://github.com/varunsridharan/wp-post/issues/).


## 📜  License & Conduct
- [**GNU General Public License v3.0**](https://github.com/varunsridharan/wp-post/blob/main/LICENSE) © [Varun Sridharan](website)
- [Code of Conduct](https://github.com/varunsridharan/.github/blob/main/CODE_OF_CONDUCT.md)


## 📣 Feedback
- ⭐ This repository if this project helped you! :wink:
- Create An [🔧 Issue](https://github.com/varunsridharan/wp-post/issues/) if you need help / found a bug


## 💰 Sponsor
[I][twitter] fell in love with open-source in 2013 and there has been no looking back since! You can read more about me [here][website].
If you, or your company, use any of my projects or like what I’m doing, kindly consider backing me. I'm in this for the long run.

- ☕ How about we get to know each other over coffee? Buy me a cup for just [**$9.99**][buymeacoffee]
- ☕️☕️ How about buying me just 2 cups of coffee each month? You can do that for as little as [**$9.99**][buymeacoffee]
- 🔰         We love bettering open-source projects. Support 1-hour of open-source maintenance for [**$24.99 one-time?**][paypal]
- 🚀         Love open-source tools? Me too! How about supporting one hour of open-source development for just [**$49.99 one-time ?**][paypal]

<!-- Personl Links -->
[paypal]: https://sva.onl/paypal
[buymeacoffee]: https://sva.onl/buymeacoffee
[twitter]: https://sva.onl/twitter/
[website]: https://sva.onl/website/


## Connect & Say 👋
- **Follow** me on [👨‍💻 Github][github] and stay updated on free and open-source software
- **Follow** me on [🐦 Twitter][twitter] to get updates on my latest open source projects
- **Message** me on [📠 Telegram][telegram]
- **Follow** my pet on [Instagram][sofythelabrador] for some _dog-tastic_ updates!

<!-- Personl Links -->
[sofythelabrador]: https://www.instagram.com/sofythelabrador/
[github]: https://sva.onl/github/
[twitter]: https://sva.onl/twitter/
[telegram]: https://sva.onl/telegram/


---

<p align="center">
<i>Built With ♥ By <a href="https://sva.onl/twitter"  target="_blank" rel="noopener noreferrer">Varun Sridharan</a> <a href="https://en.wikipedia.org/wiki/India">
   <img src="https://cdn.svarun.dev/flag-india.jpg" width="20px"/></a> </i> <br/><br/>
   <img src="https://cdn.svarun.dev/codeispoetry.png"/>
</p>

---


<!-- END common-footer.mustache  -->


CHANGELOG.md
# 📝  Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.6 - 11/04/2021
### Changed
* Codeclean up

## 1.5 - 07/05/2019
### Added
* `class_exists` Check in php

## 1.4 - 07/05/2019
### Changed
* Minor Update

## 1.3 - 14/01/2019
### Fixed
* Issue With Autoloader


## 1.2 - 20/12/2018
### Added
* `composer.json` With PSR4 Standards Autoloader

## 1.1 - 20/12/2018
### Added
* `Composer.json` Auto load Support

## 1.0 - 17/12/2018
### First Release


<!--
## Unreleased

## 1.0 - 01/02/2020
### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

-->


class-post.php
<?php
/**
 * Simple WP Post Class With Advanced Options.
 *
 * @author    Varun Sridharan <varunsridharan23@gmail.com>
 * @copyright 2018 Varun Sridharan
 * @license   GPLV3 Or Greater
 */

namespace Varunsridharan\WordPress;

/**
 * Class Post
 *
 * @package Varunsridharan\WordPress
 * @author Varun Sridharan <varunsridharan23@gmail.com>
 * @since 1.0
 */
class Post {
	/**
	 * Post Version
	 *
	 * @var string
	 */
	public $version = '1.0';

	/**
	 * Post id
	 *
	 * @var null
	 */
	protected $id = null;

	/**
	 * Post Object.
	 *
	 * @var null|\WP_Post
	 */
	protected $post = null;

	/**
	 * Returns All Taxonomies.
	 *
	 * @var array
	 */
	protected $taxes = array();

	/**
	 * All Post Data.
	 *
	 * @var array
	 */
	protected $data = array();


	/**
	 * VS_WP_Post constructor.
	 *
	 * @param int|string $post
	 */
	public function __construct( $post = '' ) {
		$this->set_post( \get_post( $post ) );
	}

	/**
	 * Sets Post Data.
	 *
	 * @param $post
	 */
	protected function set_post( $post ) {
		$this->post = $post;
		$this->id   = ( isset( $post->ID ) ) ? $post->ID : false;

		if ( ! empty( $this->taxonomies() ) ) {
			foreach ( $this->taxonomies() as $taxonomy ) {
				$this->{$taxonomy} = $this->get_terms( $taxonomy, array( 'fields' => 'all' ) );
			}
		}
	}

	/**
	 * @param $name
	 *
	 * @return bool|mixed
	 */
	public function __get( $name ) {
		if ( isset( $this->data[ $name ] ) ) {
			return $this->data[ $name ];
		}
		return false;
	}

	/**
	 * Sets A Value to $data array
	 *
	 * @uses \VS_WP_Post::$data
	 *
	 * @param $name
	 * @param $value
	 */
	public function __set( $name, $value ) {
		$this->data[ $name ] = $value;
	}

	/**
	 * Returns Post ID.
	 *
	 * @return null
	 */
	public function id() {
		return $this->id;
	}

	/**
	 * Returns Post Object.
	 *
	 * @return null
	 */
	public function post() {
		return $this->post;
	}

	/**
	 * Checks If Post Has Featured Image.
	 *
	 * @return bool
	 */
	public function has_featured_image() {
		return ( \get_the_post_thumbnail_url( $this->id() ) !== false ) ? true : false;
	}

	/**
	 * Returns Featured Image ID.
	 *
	 * @return bool|int|mixed|string
	 */
	public function featured_image_id() {
		if ( isset( $this->featured_image_id ) ) {
			return $this->featured_image_id;
		}
		$this->featured_image_id = \get_post_thumbnail_id( $this->post );
		return $this->featured_image_id;
	}

	/**
	 * Returns Featured Image URL.
	 *
	 * @param string $size
	 * @param string $default
	 *
	 * @return false|string
	 */
	public function featured_image_url( $size = 'thumbnail', $default = '' ) {
		$featured_image = \get_the_post_thumbnail_url( $this->id(), $size );
		return ( false !== $featured_image ) ? $featured_image : $default;
	}

	/**
	 * Returns Post Permalink.
	 *
	 * @return false|string
	 */
	public function permalink() {
		return \get_permalink( $this->id() );
	}

	/**
	 * Returns Post Author.
	 *
	 * @param bool|boolean $only_id
	 *
	 * @return bool
	 */
	public function author( $only_id = true ) {
		if ( true === $only_id ) {
			return $this->post->post_author;
		}
		return false;
	}

	/**
	 * Checks if post has content.
	 *
	 * @return bool
	 */
	public function has_content() {
		return ( empty( $this->post->post_content ) ) ? false : true;
	}

	/**
	 * Checks if post has post_excerpt.
	 *
	 * @return bool
	 */
	public function has_excerpt() {
		return ( empty( $this->post->has_excerpt ) ) ? false : true;
	}

	/**
	 * Returns Post Title.
	 *
	 * @return mixed
	 */
	public function title() {
		return $this->post->post_title;
	}

	/**
	 * Returns Post Slug.
	 *
	 * @return string
	 */
	public function slug() {
		return $this->post->post_name;
	}

	/**
	 * Returns Post Parent.
	 *
	 * @param boolean $only_id
	 *
	 * @return int
	 */
	public function parent( $only_id = true ) {
		if ( $only_id ) {
			return $this->post->post_parent;
		}
		$class = get_called_class();
		return new $class( $this->post->post_parent );
	}

	/**
	 * Returns Current Post Type.
	 *
	 * @return string
	 */
	public function type() {
		return $this->post->post_type;
	}

	/**
	 * Returns Page Template.
	 *
	 * @return string
	 */
	public function page_template() {
		return $this->post->page_template;
	}

	/**
	 * Returns Post post_excerpt.
	 *
	 * @return mixed
	 */
	public function excerpt() {
		return $this->post->post_excerpt;
	}

	/**
	 * Returns Post Content.
	 *
	 * @return string
	 */
	public function content() {
		return $this->post->post_content;
	}

	/**
	 * Returns Post Status.
	 *
	 * @return mixed
	 */
	public function status() {
		return $this->post->post_status;
	}

	/**
	 * Checks if given status is post status.
	 *
	 * @param string $status
	 *
	 * @return bool
	 */
	public function is_status( $status = 'publish' ) {
		return ( $status === $this->post->post_status ) ? true : false;
	}

	/*====================================================================================
	 * Post Meta Handler.																 *
	 ====================================================================================*/

	/**
	 * Gets Post Meta For the given meta key.
	 *
	 * @param string $meta_key
	 * @param string $default
	 *
	 * @return mixed
	 */
	public function get_meta( $meta_key = '', $default = '' ) {
		return \get_post_meta( $this->id(), $meta_key, $default );
	}

	/**
	 * Updates Post Meta.
	 *
	 * @param string $meta_key
	 * @param string $values
	 * @param string $prev_values
	 *
	 * @return bool|int
	 */
	public function update_meta( $meta_key = '', $values = '', $prev_values = '' ) {
		return \update_post_meta( $this->id(), $meta_key, $values, $prev_values );
	}

	/**
	 * Adds Post Meta.
	 *
	 * @param string       $meta_key
	 * @param string       $values
	 * @param boolean|bool $unique
	 *
	 * @return bool|int
	 */
	public function add_meta( $meta_key = '', $values = '', $unique = false ) {
		return \add_post_meta( $this->id(), $meta_key, $values, $unique );
	}

	/**
	 * Deletes A Post Meta.
	 *
	 * @param string $meta_key
	 * @param mixed  $values
	 *
	 * @return bool
	 */
	public function delete_meta( $meta_key = '', $values = '' ) {
		return \delete_post_meta( $this->id(), $meta_key, $values );
	}

	/*====================================================================================
	 * Taxonomy Handler.																 *
	 ====================================================================================*/

	/**
	 * Returns All Post Taxonomies.
	 *
	 * @return array
	 */
	public function taxonomies() {
		if ( ! empty( $this->taxes ) ) {
			return $this->taxes;
		}
		$taxes       = \get_post_taxonomies( $this->post );
		$this->taxes = $taxes;
		return $this->taxes;
	}

	/**
	 * Returns All Terms From Post.
	 *
	 * @param string $taxonomy
	 * @param array  $args
	 *
	 * @return array|\WP_Error
	 */
	public function get_terms( $taxonomy = '', $args = array() ) {
		if ( isset( $this->{$taxonomy} ) && empty( $args ) ) {
			return $this->{$taxonomy};
		} else {
			$name = $taxonomy . '_' . md5( json_encode( $args ) );
			if ( isset( $this->{$name} ) ) {
				return $this->{$name};
			}
			$data          = \wp_get_post_terms( $this->id(), $taxonomy, $args );
			$this->{$name} = $data;
			return $data;
		}
	}

	/**
	 * Sets Given Terms With Tax To The Current Post.
	 *
	 * @param array  $terms
	 * @param string $taxonomy
	 * @param bool   $append
	 *
	 * @return array|false|\WP_Error
	 */
	public function set_terms( $terms = array(), $taxonomy = '', $append = false ) {
		return \wp_set_post_terms( $this->id(), $terms, $taxonomy, $append );
	}


	/*====================================================================================
	 * Class Helpers.
	 ====================================================================================*/

	/**
	 * Checks if given key is in array
	 *
	 * @uses \in_array()
	 * @uses \VS_WP_Post
	 *
	 * @example in('somekey','post_category','no') post_category will be taken from post data.
	 * @example in('somekey',array('somekey' => 'somekey'), will be used the given array.
	 *
	 * @param string $key
	 * @param string $data
	 * @param bool   $default
	 *
	 * @return bool|string
	 */
	public function in( $key = '', $data = '', $default = false ) {
		if ( is_array( $data ) ) {
			return ( in_array( $key, $data ) ) ? $key : $default;
		}
		if ( isset( $this->{$data} ) ) {
			return ( in_array( $key, $this->{$data} ) ) ? $key : $default;
		}
		return $default;
	}
}



class-wp-post.php
<?php
/**
 * VSP WP Post Helper
 *
 * User: varun
 * Date: 01-04-2018
 * Time: 07:22 AM
 *
 * @author    Varun Sridharan <varunsridharan23@gmail.com>
 * @since     1.0
 * @version   1.0
 * @copyright GPL V3 Or greater
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

if ( ! class_exists( 'VS_WP_Post' ) ) {
	/**
	 * Class VS_WP_Post
	 */
	class VS_WP_Post {
		/**
		 * VS_WP_Post Version
		 *
		 * @var string
		 */
		public $version = '1.0';

		/**
		 * Post id
		 *
		 * @var null
		 */
		protected $id = null;

		/**
		 * Post Object.
		 *
		 * @var null|\WP_Post
		 */
		protected $post = null;

		/**
		 * Returns All Taxonomies.
		 *
		 * @var array
		 */
		protected $taxes = array();

		/**
		 * All Post Data.
		 *
		 * @var array
		 */
		protected $data = array();


		/**
		 * VS_WP_Post constructor.
		 *
		 * @param int|string $post
		 */
		public function __construct( $post = '' ) {
			$this->set_post( get_post( $post ) );
		}

		/**
		 * Sets Post Data.
		 *
		 * @param $post
		 */
		protected function set_post( $post ) {
			$this->post = $post;
			$this->id   = ( isset( $post->ID ) ) ? $post->ID : false;

			if ( ! empty( $this->taxonomies() ) ) {
				foreach ( $this->taxonomies() as $taxonomy ) {
					$this->{$taxonomy} = $this->get_terms( $taxonomy, array( 'fields' => 'all' ) );
				}
			}
		}

		/**
		 * @param $name
		 *
		 * @return bool|mixed
		 */
		public function __get( $name ) {
			if ( isset( $this->data[ $name ] ) ) {
				return $this->data[ $name ];
			}
			return false;
		}

		/**
		 * Sets A Value to $data array
		 *
		 * @uses \VS_WP_Post::$data
		 *
		 * @param $name
		 * @param $value
		 */
		public function __set( $name, $value ) {
			$this->data[ $name ] = $value;
		}

		/**
		 * Returns Post ID.
		 *
		 * @return null
		 */
		public function id() {
			return $this->id;
		}

		/**
		 * Returns Post Object.
		 *
		 * @return null
		 */
		public function post() {
			return $this->post;
		}

		/**
		 * Checks If Post Has Featured Image.
		 *
		 * @return bool
		 */
		public function has_featured_image() {
			return ( get_the_post_thumbnail_url( $this->id() ) !== false ) ? true : false;
		}

		/**
		 * Returns Featured Image ID.
		 *
		 * @return bool|int|mixed|string
		 */
		public function featured_image_id() {
			if ( isset( $this->featured_image_id ) ) {
				return $this->featured_image_id;
			}
			$this->featured_image_id = get_post_thumbnail_id( $this->post );
			return $this->featured_image_id;
		}

		/**
		 * Returns Featured Image URL.
		 *
		 * @param string $size
		 * @param string $default
		 *
		 * @return false|string
		 */
		public function featured_image_url( $size = 'thumbnail', $default = '' ) {
			$featured_image = get_the_post_thumbnail_url( $this->id(), $size );
			return ( false !== $featured_image ) ? $featured_image : $default;
		}

		/**
		 * Returns Post Permalink.
		 *
		 * @return false|string
		 */
		public function permalink() {
			return get_permalink( $this->id() );
		}

		/**
		 * Returns Post Author.
		 *
		 * @param bool|boolean $only_id
		 *
		 * @return bool
		 */
		public function author( $only_id = true ) {
			if ( true === $only_id ) {
				return $this->post->post_author;
			}
			return false;
		}

		/**
		 * Checks if post has content.
		 *
		 * @return bool
		 */
		public function has_content() {
			return ( empty( $this->post->post_content ) ) ? false : true;
		}

		/**
		 * Checks if post has post_excerpt.
		 *
		 * @return bool
		 */
		public function has_excerpt() {
			return ( empty( $this->post->has_excerpt ) ) ? false : true;
		}

		/**
		 * Returns Post Title.
		 *
		 * @return mixed
		 */
		public function title() {
			return $this->post->post_title;
		}

		/**
		 * Returns Post Slug.
		 *
		 * @return string
		 */
		public function slug() {
			return $this->post->post_name;
		}

		/**
		 * Returns Post Parent.
		 *
		 * @param boolean $only_id
		 *
		 * @return int
		 */
		public function parent( $only_id = true ) {
			if ( $only_id ) {
				return $this->post->post_parent;
			}
			$class = get_called_class();
			return new $class( $this->post->post_parent );
		}

		/**
		 * Returns Current Post Type.
		 *
		 * @return string
		 */
		public function type() {
			return $this->post->post_type;
		}

		/**
		 * Returns Page Template.
		 *
		 * @return string
		 */
		public function page_template() {
			return $this->post->page_template;
		}

		/**
		 * Returns Post post_excerpt.
		 *
		 * @return mixed
		 */
		public function excerpt() {
			return $this->post->post_excerpt;
		}

		/**
		 * Returns Post Content.
		 *
		 * @return string
		 */
		public function content() {
			return $this->post->post_content;
		}

		/**
		 * Returns Post Status.
		 *
		 * @return mixed
		 */
		public function status() {
			return $this->post->post_status;
		}

		/**
		 * Checks if given status is post status.
		 *
		 * @param string $status
		 *
		 * @return bool
		 */
		public function is_status( $status = 'publish' ) {
			return ( $status === $this->post->post_status ) ? true : false;
		}

		/*====================================================================================
		 * Post Meta Handler.																 *
		 ====================================================================================*/

		/**
		 * Gets Post Meta For the given meta key.
		 *
		 * @param string $meta_key
		 * @param string $default
		 *
		 * @return mixed
		 */
		public function get_meta( $meta_key = '', $default = '' ) {
			return get_post_meta( $this->id(), $meta_key, $default );
		}

		/**
		 * Updates Post Meta.
		 *
		 * @param string $meta_key
		 * @param string $values
		 * @param string $prev_values
		 *
		 * @return bool|int
		 */
		public function update_meta( $meta_key = '', $values = '', $prev_values = '' ) {
			return update_post_meta( $this->id(), $meta_key, $values, $prev_values );
		}

		/**
		 * Adds Post Meta.
		 *
		 * @param string       $meta_key
		 * @param string       $values
		 * @param boolean|bool $unique
		 *
		 * @return bool|int
		 */
		public function add_meta( $meta_key = '', $values = '', $unique = false ) {
			return add_post_meta( $this->id(), $meta_key, $values, $unique );
		}

		/**
		 * Deletes A Post Meta.
		 *
		 * @param string $meta_key
		 * @param mixed  $values
		 *
		 * @return bool
		 */
		public function delete_meta( $meta_key = '', $values = '' ) {
			return delete_post_meta( $this->id(), $meta_key, $values );
		}

		/*====================================================================================
		 * Taxonomy Handler.																 *
		 ====================================================================================*/

		/**
		 * Returns All Post Taxonomies.
		 *
		 * @return array
		 */
		public function taxonomies() {
			if ( ! empty( $this->taxes ) ) {
				return $this->taxes;
			}
			$taxes       = get_post_taxonomies( $this->post );
			$this->taxes = $taxes;
			return $this->taxes;
		}

		/**
		 * Returns All Terms From Post.
		 *
		 * @param string $taxonomy
		 * @param array  $args
		 *
		 * @return array|\WP_Error
		 */
		public function get_terms( $taxonomy = '', $args = array() ) {
			if ( isset( $this->{$taxonomy} ) && empty( $args ) ) {
				return $this->{$taxonomy};
			} else {
				$name = $taxonomy . '_' . md5( json_encode( $args ) );
				if ( isset( $this->{$name} ) ) {
					return $this->{$name};
				}
				$data          = wp_get_post_terms( $this->id(), $taxonomy, $args );
				$this->{$name} = $data;
				return $data;
			}
		}

		/**
		 * Sets Given Terms With Tax To The Current Post.
		 *
		 * @param array  $terms
		 * @param string $taxonomy
		 * @param bool   $append
		 *
		 * @return array|false|\WP_Error
		 */
		public function set_terms( $terms = array(), $taxonomy = '', $append = false ) {
			return wp_set_post_terms( $this->id(), $terms, $taxonomy, $append );
		}


		/*====================================================================================
		 * Class Helpers.
		 ====================================================================================*/

		/**
		 * Checks if given key is in array
		 *
		 * @uses \in_array()
		 * @uses \VS_WP_Post
		 *
		 * @example in('somekey','post_category','no') post_category will be taken from post data.
		 * @example in('somekey',array('somekey' => 'somekey'), will be used the given array.
		 *
		 * @param string $key
		 * @param string $data
		 * @param bool   $default
		 *
		 * @return bool|string
		 */
		public function in( $key = '', $data = '', $default = false ) {
			if ( is_array( $data ) ) {
				return ( in_array( $key, $data ) ) ? $key : $default;
			}
			if ( isset( $this->{$data} ) ) {
				return ( in_array( $key, $this->{$data} ) ) ? $key : $default;
			}
			return $default;
		}
	}
}
